doc/generic/pgf/text-en/pgfmanual-en-base-shadings.tex

```diff
@@ -313,9 +395,184 @@ \subsubsection{General (Functional) Shadings}
+        \meta{color name}.
+        %
+    \end{command}
+
+    \begin{command}{\pgfshadecolortogray\marg{color name}\marg{macro}}
+        This command takes \meta{color name} as input converts it to grayscale
+        and stores the color's value as a real number between 0.0 and 1.0.
+
+        Although it's not needed, for consistency a second macro suffixed with
+        |gray| is also defined.
+        %
+    \end{command}
     %
 \end{command}
 
+\paragraph{Color model independent functional shadings.}
+
+By nature, the PostScript code used in functional shadings must output one of
+\textsc{rgb}, \textsc{cmyk}, or grayscale data. Therefore,
+\cs{pgfdeclarefunctionalshading} is \emph{not} portable across color models.
+
+Take particular care that the same color model is in use at declaration time
+and use time for functional shadings declared with an optional argument as
+otherwise the PostScript data will not match the declared color space and
+you will end up with a malformed PDF.
+
+Having said this, it \emph{is} possible to create portable functional shadings
+by providing conditional code to append color transformations to the
+PostScript data. A variety of \cs{pgffuncshading*to*} (e.g.,
+\cs{pgffuncshadingrgbtocmyk}) macros along with \cs{ifpgfshadingmodel*} (e.g.,
+\cs{ifpgfshadingmodelcmyk}) conditionals are provided to assist with these
+transformations. Obviously, this will make the PostScript code less efficient
+than if you work in your intended color model.
+
+\pgfdeclarefunctionalshading[black]{portabletwospots}
+    {\pgfpointorigin}{\pgfpoint{3.5cm}{3.5cm}}{}{
+  2 copy
+  45 sub dup mul exch
+  40 sub dup mul 0.5 mul add sqrt
+  dup mul neg 1.0005 exch exp 1.0 exch sub
+  3 1 roll
+  70 sub dup mul .5 mul exch
+  70 sub dup mul add sqrt
+  dup mul neg 1.002 exch exp 1.0 exch sub
+  1.0 3 1 roll
+  \ifpgfshadingmodelcmyk
+    \pgffuncshadingrgbtocmyk
+  \fi
+  \ifpgfshadingmodelgray
+    \pgffuncshadingrgbtogray
+  \fi
+}
+\begin{center}
+  \begin{minipage}{3.5cm}
+    |xcolor| |rgb| model:\medskip
+
+    \selectcolormodel{rgb}
+    \pgfuseshading{portabletwospots}
+  \end{minipage}\hspace{2cm}
+  \begin{minipage}{3.5cm}
+    |xcolor| |cmyk| model:\medskip
+
+    \selectcolormodel{cmyk}
+    \pgfuseshading{portabletwospots}
+  \end{minipage}\hspace{2cm}
+  \begin{minipage}{3.5cm}
+    |xcolor| |gray| model:\medskip
+
+    \selectcolormodel{gray}
+    \pgfuseshading{portabletwospots}
+  \end{minipage}
+\end{center}
+
+\begin{codeexample}[code only]
+\pgfdeclarefunctionalshading[black]{portabletwospots}{\pgfpointorigin}{\pgfpoint{3.5cm}{3.5cm}}{}{
+  2 copy
+  45 sub dup mul exch
+  40 sub dup mul 0.5 mul add sqrt
+  dup mul neg 1.0005 exch exp 1.0 exch sub
+  3 1 roll
+  70 sub dup mul .5 mul exch
+  70 sub dup mul add sqrt
+  dup mul neg 1.002 exch exp 1.0 exch sub
+  1.0 3 1 roll
+  \ifpgfshadingmodelcmyk
+    \pgffuncshadingrgbtocmyk
+  \fi
+  \ifpgfshadingmodelgray
+    \pgffuncshadingrgbtogray
+  \fi
+}
```

Comment:
Do you think it would make sense to just automatically append the bit
```latex
  \ifpgfshadingmodelcmyk
    \pgffuncshadingrgbtocmyk
  \fi
  \ifpgfshadingmodelgray
    \pgffuncshadingrgbtogray
  \fi
```
to `\pgfdeclarefunctionalshading`?

Reply:
Yes, this could make sense and I'm happy to adjust it, but there would need to be a package option (or some other flag) to enable or disable automatic conversion.

The disadvantage of making it automatic (with no option) is that it forces you to always work in RGB. But the conversion from RGB to CMYK makes assumptions regarding under colour removal and black generation that you might not want if you were working in CMYK directly. Since there are no colour profiles involved, I think it is important to be able to work in your intended output colour space directly.

So a few of questions:

1. Do you have a preference for how this option would be controlled? (e.g., a package option, a Boolean somewhere, something else, etc.)
2. Could `\ifpgfshadingmodel*` and `\pgffuncshadingrgbto*` then become private and removed from the manual?
2. Are the `\pgffuncshadingcmykto*` and `\pgffuncshadinggrayto*` macros needed at all?

Reply:
I think this is fine for now and I will just merge it.  Before making a new release we can think about it again.  I don't think `\pgfdeclarefunctionalshading` is widely used.